### ISSUE TITLE
robustly close files in FilePath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4400]
|<img alt="benjamin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| benjamin reported|
|-|-|
|Trac ID|trac#4400|
|Type|enhancement|
|Created|2010-04-15 02:44:17Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-close-4400|

This patch uses try, finally to ensure files are closed. It also prevents ensures proper file closing on non-refcounting gcs.

Attachments:

* [proper_file_closing.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/543/543832b268a002be6f1c16455767ee68f44853ca/344fa83869713a282c79783ad6aae7792fbf04bf.patch) (2570 bytes) - added by benjamin on 2010-04-15 02:44:43Z - 
* [proper_file_closing_pendantic_testing.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/543/543832b268a002be6f1c16455767ee68f44853ca/a9ab7e0d4af6e7b87fe9101ad18d4a6645e3c865.patch) (4589 bytes) - added by benjamin on 2010-04-15 20:29:03Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4400 4400
type__enhancement enhancement
reporter__benjamin benjamin
priority__normal normal
milestone__ 
branch__branches_filepath_close_4400 branches/filepath-close-4400
branch_author__benjamin benjamin
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1271299457000000 1271299457000000
changetime__1271818935000000 1271818935000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to benjamin|
|-|-|

Hi benjamin!  Thanks for the patch!

In order to apply this, we need some unit tests which verify that the problem was actually fixed.  While you can't change the reference-counting GC, you can temporarily stub out the implementation of 'open' used by FilePath so that something holds on to a reference to the underlying file.

Could you add some such tests to the patch and re-attach the 'review' keyword?

**Comment by twisted-trac** [comment]
|<img alt="benjamin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|benjamin removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [28833]) Branching to 'filepath-close-4400'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks for the updated patch with unit tests.  The change looks correct and the test coverage seems complete.  I do have have two other comments about the updated patch though:

  * Everything should have docstrings, even test helpers, to help out future maintainers.  So all of `ExplodingFile`'s methods, and the class itself, should be documented.  Likewise for `ExplodingFilePath`.  Even more importantly, the test methods should have docstrings explaining what they're testing.
  * It's better if each test method only tests one thing.  Lots of existing `FilePath` tests put multiple tests into a single method, so I'm not surprised you did it this way. :)  New tests shouldn't follow this convention, though.

These are both very simple to fix, so I think I'll do that and then apply the change.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [28835]) Docstrings for ExplodingFile

refs [#4400](https://github.com/twisted/twisted/issues/4400)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [28836]) Docstrings for ExplodingFilePath

refs [#4400](https://github.com/twisted/twisted/issues/4400)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [28837]) Split test_getAndSetFileClosing into two methods; add docstrings for both

refs [#4400](https://github.com/twisted/twisted/issues/4400)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [28838]) Remove the duplicate assertions from `test_copyToFileClosing` and give it a docstring and a couple nicer variable names

refs [#4400](https://github.com/twisted/twisted/issues/4400)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [28839]) news fragment

refs [#4400](https://github.com/twisted/twisted/issues/4400)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [28840]) Restore the test for proper closing if the open fails

refs [#4400](https://github.com/twisted/twisted/issues/4400)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [28841]) Merge filepath-close-4400

Author: benjamin, exarkun
Reviewer: exarkun
Fixes: [#4400](https://github.com/twisted/twisted/issues/4400)

Add try/finally statements to `FilePath` around code that opens files to make
sure that those files get closed, even if an exception is raised by some I/O
operation.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
